sourcegraph/cody
# [feat] Token usage indicator in the chat view

## Description

This PR introduces a new feature that displays the remaining token counts for chat, user, and enhanced token types in the chat view. This enhancement aims to provide users with real-time visibility into their token usage, helping them manage their token allocation more effectively.

## Changes

- Updated the `TokenCounter` class to expose a new method, `getRemainingTokens()`, which returns an object containing the remaining token counts for each token type.
- Modified the `SimpleChatPanelProvider` to:
  - Retrieve the remaining token counts from the `TokenCounter` instance after building the prompt.
  - Send the remaining token counts to the webview using the `postMessage` method.
- Updated the prompt building logic to return both the `PromptInfo` and `TokenCounter` instances.
- Added a new `remainingTokens` message type to the extension-webview communication protocol to facilitate the transfer of token counts.
- Created a new `TokenIndicators` component that displays the remaining token counts in the chat view UI.
- Integrated the `TokenIndicators` component into the `Chat` component, populating it with the token counts received from the extension.
- Added styling for the `TokenIndicators` component to ensure a visually appealing and intuitive user experience.

## Benefits

- Users can now see their remaining token counts for each token type directly in the chat view.
- The real-time display of token usage helps users make informed decisions about their token allocation and usage.
- The integration of token indicators enhances the overall user experience and provides valuable insights into token consumption.

## Screenshots

![Screenshot 2024-06-06 095520](https://github.com/sourcegraph/cody/assets/31413214/ca3b46d4-bb9b-425e-a716-da1d245220dd)


## Checklist

- [x] Tested the changes manually
- [x] Followed the project's coding conventions and style guidelines
- [x] Wrote clear and concise commit messages following the conventional commit format

Please review this PR and provide any feedback or suggestions for improvement. Let me know if you have any questions or concerns.


## Test plan

- Manually tested the feature by engaging in chat conversations and verifying that the token indicators update accurately based on the token usage.
- Ensured that the token counts are correctly retrieved from the `TokenCounter` and propagated to the webview.
- Verified that the `TokenIndicators` component renders properly and displays the expected token counts.

## How to test

1. Build extension from this branch
2. Open a new Chat
3. Ask question w/ or w/o enhanced context
4. Check how the Remaining Token UI changes.
5. Choose different models to see usage changes.
